File: apps/ios/Barns/BarnsApp/Presentation/MyItems/ArchivedGreeneryViewModel.swift
import Combine

/// Drives the local-only Archived Greenery list. It loads the customer's
/// archived `ProductItem`s (read-only here, except restore) and can restore an
/// item back to the active My Greenery list. All work is local: no hard delete,
/// no API, no sync.
@MainActor
final class ArchivedGreeneryViewModel: ObservableObject {
    enum State: Equatable {
        case loading
        case loaded([ProductItem])
        case failed(String)
    }

    @Published private(set) var state: State = .loading

    private let getProductItemsUseCase: GetProductItemsUseCase
    private let restoreProductItemUseCase: RestoreProductItemUseCase

    init(
        getProductItemsUseCase: GetProductItemsUseCase,
        restoreProductItemUseCase: RestoreProductItemUseCase
    ) {
        self.getProductItemsUseCase = getProductItemsUseCase
        self.restoreProductItemUseCase = restoreProductItemUseCase
    }

    func load() async {
        state = .loading
        do {
            let items = try await getProductItemsUseCase.execute()
            // This list shows only archived items; they remain in the local
            // store (no hard delete).
            state = .loaded(items.filter { $0.status == .archived })
        } catch {
            state = .failed("Unable to load archived items. Please try again.")
        }
    }

    /// Restores an archived item back to active locally (soft action; no hard
    /// delete). Reloads so the restored item leaves this archived list.
    /// Returns true on success.
    @discardableResult
    func restore(_ item: ProductItem) async -> Bool {
        do {
            _ = try await restoreProductItemUseCase.execute(item)
            await load()
            return true
        } catch {
            state = .failed("Unable to restore this item. Please try again.")
            return false
        }
    }
}

File: apps/android/app/src/main/java/com/barns/app/presentation/myitems/ArchivedGreeneryViewModel.kt
package com.barns.app.presentation.myitems

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.barns.app.domain.model.ProductItem
import com.barns.app.domain.model.ProductItemStatus
import com.barns.app.domain.usecase.myitems.GetProductItemsUseCase
import com.barns.app.domain.usecase.myitems.RestoreProductItemUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Drives the local-only Archived Greenery list. It loads the customer's
 * archived [ProductItem]s (read-only here, except restore) and can restore an
 * item back to the active My Greenery list. All work is local: no hard delete,
 * no re-create/re-insert, no API, no sync.
 */
class ArchivedGreeneryViewModel(
    private val getProductItemsUseCase: GetProductItemsUseCase,
    private val restoreProductItemUseCase: RestoreProductItemUseCase,
) : ViewModel() {
    sealed interface State {
        object Loading : State
        data class Loaded(val items: List<ProductItem>) : State
        data class Failed(val message: String) : State
    }

    private val _state = MutableStateFlow<State>(State.Loading)
    val state: StateFlow<State> = _state.asStateFlow()

    fun load() {
        _state.value = State.Loading
        viewModelScope.launch {
            runCatching { getProductItemsUseCase.execute() }
                // This list shows only archived items; they remain in the local
                // store (no hard delete).
                .onSuccess { items ->
                    _state.value = State.Loaded(items.filter { it.status == ProductItemStatus.ARCHIVED })
                }
                .onFailure {
                    _state.value = State.Failed("Unable to load archived items. Please try again.")
                }
        }
    }

    /**
     * Restores an archived item back to active locally (soft action; no hard
     * delete). Reloads so the restored item leaves this archived list.
     */
    fun restore(item: ProductItem) {
        viewModelScope.launch {
            runCatching { restoreProductItemUseCase.execute(item) }
                .onSuccess { load() }
                .onFailure {
                    _state.value = State.Failed("Unable to restore this item. Please try again.")
                }
        }
    }
}
